### PR TITLE
Resolve crash when editor tab is closed.

### LINKED ISSUE #2758: Deleted C++ object: WebEngineView when closing editor after closing tab
When opening an editor for a form, then closing the tab, and then the editor, we get:

```
$ python3 -m qutebrowser --temp-basedir --backend webengine --debug tests/end2end/data/hello.txt tests/end2end/data/insert_mode_settings/html/textarea.html
[...]
19:40:01 DEBUG    commands   command:run:524 Calling qutebrowser.browser.commands.CommandDispatcher.open_editor(<qutebrowser.browser.commands.CommandDispatcher>)
19:40:01 DEBUG    webview    webenginetab:_js_cb_single:494 Got element from JS: {'attributes': {'id': 'qute-textarea'}, 'class_name': '', 'id': 1, 'outer_xml': '<textarea id="qute-textarea"></textarea>', 'rects': [{'bottom': 44, 'height': 36, 'left': 8, 'right': 189, 'top': 8, 'width': 181}], 'tag_name': 'TEXTAREA', 'text': '', 'value': ''}
19:40:01 DEBUG    webelem    webelem:is_editable:238 Checking if element is editable: <qutebrowser.browser.webengine.webengineelem.WebEngineElement html='<textarea id="qute-textarea"></textarea>'>
19:40:01 DEBUG    procs      editor:edit:126 Calling "gvim" with args ['-f', '/tmp/qutebrowser-editor-rp75_n9j']
19:40:01 DEBUG    procs      guiprocess:start:139 Starting process.
19:40:01 DEBUG    procs      guiprocess:_pre_start:133 Executing: gvim -f /tmp/qutebrowser-editor-rp75_n9j
19:40:01 DEBUG    procs      guiprocess:on_started:122 Process started.
[...]
19:40:04 DEBUG    commands   command:run:524 Calling qutebrowser.browser.commands.CommandDispatcher.tab_close(<qutebrowser.browser.commands.CommandDispatcher>, False, False, False, False, None)
[...]
19:40:06 DEBUG    procs      guiprocess:on_finished:98 Process finished with code 0, status 0.
19:40:06 DEBUG    procs      editor:on_proc_closed:70 Editor closed
19:40:06 DEBUG    procs      editor:on_proc_closed:87 Read back: 
19:40:06 ERROR    misc       crashsignal:exception_hook:205 Uncaught exception
Traceback (most recent call last):
  File "/home/florian/proj/qutebrowser/git/qutebrowser/browser/commands.py", line 1617, in on_editing_finished
    elem.set_value(text)
  File "/home/florian/proj/qutebrowser/git/qutebrowser/browser/webengine/webengineelem.py", line 133, in set_value
    self._js_call('set_value', value)
  File "/home/florian/proj/qutebrowser/git/qutebrowser/browser/webengine/webengineelem.py", line 103, in _js_call
    self._tab.run_js_async(js_code, callback=callback)
  File "/home/florian/proj/qutebrowser/git/qutebrowser/browser/webengine/webenginetab.py", line 611, in run_js_async
    self._widget.page().runJavaScript(code, world_id)
RuntimeError: wrapped C/C++ object of type WebEngineView has been deleted
```

### PR DESCRIPTION
**Pull request by rcorre**
If an editor is open on a form in a tab and that tab is closed, rewire
the callback to print a warning. Previously, the callback would access a
deleted C++ object and cause a crash.
Resolves #2758.

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/3333)

### CODE REVIEW

**Comment by rcorre** [comment]
I decided not to close the editor from our end -- the user might not want to lose what they typed.

**Comment by rcorre** [comment]
Related: should `on_error` skip `cleanup` and instead print `Editor error. Text left in /tmp/qutebrowser-editor-SUFFIX` (for all errors, not just this one)?

File: qutebrowser/browser/commands.py [hunk 1]
```diff
@@ -1647,6 +1650,11 @@ def on_editing_finished(self, elem, text):
         except webelem.Error as e:
             raise cmdexc.CommandError(str(e))
 
+    def on_editor_orphaned(self, ed):
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: Hmm, while the reconnecting works, I wonder whether there isn't a better solution. Can't we detect whether the view is gone in `on_editing_finished` instead?

File: qutebrowser/browser/commands.py [hunk 2]
```diff
@@ -1647,6 +1650,11 @@ def on_editing_finished(self, elem, text):
...
+        ed.editing_finished.disconnect()
+        ed.editing_finished.connect(
+            lambda: message.warning('Edited element was closed'))
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: I'd prefer "Edited element vanished" or so (we already use that terminology elsewhere).

File: tests/end2end/features/test_editor_bdd.py [hunk 3]
```diff
@@ -64,3 +64,21 @@ def set_up_editor(quteproc, server, tmpdir, text):
...
+@bdd.when(bdd.parsers.parse(
+    'I set up a fake editor returning "{text}" after {t}s'))
+def set_up_editor_delay(quteproc, server, tmpdir, text, t):
+    """Set up editor.command to a small python script inserting a text."""
+    script = tmpdir / 'script.py'
+    script.write(textwrap.dedent("""
+        import sys
+        import time
+
+        time.sleep({t})
```
Comment (The-Compiler) [inline tests/end2end/features/test_editor_bdd.py]: This sounds like a recipe for flaky tests - can't we save the editor somewhere, and have another `@bdd.when` function quitting the editor somehow (e.g. by making the Python script exit cleanly on `SIGUSR1` or so)?

**Comment by The-Compiler** [comment]
> Related: should `on_error` skip `cleanup` and instead print `Editor error. Text left in /tmp/qutebrowser-editor-SUFFIX` (for all errors, not just this one)?

Yes, but that'd probably be a separate PR? See #1596 for that, and #1677 for an abandoned PR related to that.

File: qutebrowser/browser/commands.py (hunk 1 above)
Comment (rcorre) [reply qutebrowser/browser/commands.py]: The C++ object wrapped by the `WebEngineView` is deleted. A few ideas I had are:

- Can a `WebEngineView` check if the C++ object is deleted?
- Does `AbstractTab` expose any way to check if it is closed?
- Can I search the `TabbedBrowser` to see if it still contains a tab I hold a reference to?

I didn't see an obvious way to do any of these, so catching the shutdown signal from the tab seemed like the best solution. Any ideas?

Comment (The-Compiler) [reply qutebrowser/browser/commands.py]: You can use `sip.isdeleted` on a PyQt object. So I guess adding a `is_deleted` method to `AbstractTab` which does `return sip.isdeleted(self._widget)` or so would be the nicest solution here - what do you think?

File: tests/end2end/features/test_editor_bdd.py [hunk 4]
```diff
@@ -64,3 +66,33 @@ def set_up_editor(quteproc, server, tmpdir, text):
...
+    """Set up editor.command to a small python script inserting a text."""
+    pidfile = tmpdir / 'editor_pid'
+    script = tmpdir / 'script.py'
+    script.write(textwrap.dedent("""
+        import os
+        import sys
+        import time
+        import signal
+
+        with open('{pidfile}', 'w') as f:
```
Comment (The-Compiler) [inline tests/end2end/features/test_editor_bdd.py]: You'll need to use `r'{pidfile}'` here so it works with backslashes on Windows.

File: tests/end2end/features/test_editor_bdd.py [hunk 5]
```diff
@@ -64,3 +66,33 @@ def set_up_editor(quteproc, server, tmpdir, text):
...
+    editor = json.dumps([sys.executable, str(script), '{}'])
+    quteproc.set_setting('editor.command', editor)
+
+
+@bdd.when(bdd.parsers.parse('I kill the waiting editor'))
+def kill_editor_wait(quteproc, server, tmpdir):
+    """Kill the waiting editor."""
+    pidfile = tmpdir / 'editor_pid'
+    pid = int(pidfile.read())
+    os.kill(pid, signal.SIGUSR1)
```
Comment (The-Compiler) [inline tests/end2end/features/test_editor_bdd.py]: Dang, pylint correctly points out that this won't work on Windows 😆 I wonder whether using `SIGTERM` instead (and handling that in the process) would work on Windows as well.

**Comment by rcorre** [comment]
> Welp, looks like it still doesn't work. I'd say back to SIGUSR1 and don't run the test on Windows then...

Just looked up "python catch sigterm on windows" and I [wish I hadn't](https://stackoverflow.com/questions/35772001/how-to-handle-the-signal-in-python-on-windows-machine#comment81619892_35792192) (though maybe `SIGINT` [would work](https://stackoverflow.com/a/26073025/1435461)?)

**Comment by The-Compiler** [comment]
Yeah, Windows is... weird. Can you push a test with `SIGINT` and keep an eye on AppVeyor? :wink:

**Comment by rcorre** [comment]
`Module 'signal' has no 'SIGUSR1' member (no-member)` :roll_eyes:
Guess I still need to use SIGINT? Or just pylint ignore that on windows somehow?

**Comment by rcorre** [comment]
Also, looks like the server/quteproc fixture isn't used in many of those steps, and it seems to work fine without it. Should server/quteproc be removed from the arg list in those cases? Or does it do some important setup/teardown behind-the-scenes?

**Comment by The-Compiler** [comment]
Huh, now AppVeyor discovered something which looks like a PyQt bug:

```
Qt exceptions in virtual methods:
________________________________________________________________________________
StopIteration
The above exception was the direct cause of the following exception:
SystemError: <class 'PyQt5.QtCore.QtMsgType'> returned a result with an error set
________________________________________________________________________________
```

I'll trigger a rerun and hope for the best....

> Or just pylint ignore that on windows somehow?

Yeah, that. The ignore you did seems fine.

> Also, looks like the server/quteproc fixture isn't used in many of those steps, and it seems to work fine without it. Should server/quteproc be removed from the arg list in those cases? Or does it do some important setup/teardown behind-the-scenes?

Hmm, no, I don't think they're needed if they aren't actually accessed, at least in the BDD steps. Do you mean in that file, or in general?

File: qutebrowser/browser/commands.py [hunk 6]
```diff
@@ -1644,6 +1644,8 @@ def on_editing_finished(self, elem, text):
         """
         try:
             elem.set_value(text)
+        except webelem.OrphanedError as e:
+            message.warning('Edited element vanished')
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: I changed this to an error, that seems more appropriate than a warning here.

**Comment by The-Compiler** [comment]
Merging this as-is, thanks! If you have some change cleaning up the unneeded fixture, I'd be happy about another PR for that, though!

**Pull request merged**